compiler/noirc_frontend/src/hir/resolution/path_resolver.rs

```diff
@@ -1,16 +1,20 @@
 use super::import::{resolve_import, ImportDirective, PathResolution, PathResolutionResult};
 use crate::ast::Path;
+use crate::node_interner::ReferenceId;
 use std::collections::BTreeMap;
 
 use crate::graph::CrateId;
 use crate::hir::def_map::{CrateDefMap, LocalModuleId, ModuleId};
 
 pub trait PathResolver {
     /// Resolve the given path returning the resolved ModuleDefId.
+    /// If `path_references` is `Some`, a `ReferenceId` for each segment in `path`
+    /// will be resolved and pushed.
     fn resolve(
         &self,
         def_maps: &BTreeMap<CrateId, CrateDefMap>,
         path: Path,
+        path_references: &mut Option<&mut Vec<ReferenceId>>,
```

Comment:
Is there a more idiomatic way to do this? I want to avoid allocating a `Vec` when not in LSP mode.

Alternatively, I tried changing this to be a callback (`dyn FnMut`) but then it told me that the trait cannot be used as an object... I'm a bit lost as I'm just beginning with Rust. Any help is appreciated! 🙏

Reply:
Searching the internet, it seems it's common to use `Option<&mut ...>` or `&mut Option<&mut ...>` for this, so maybe it's fine 🤷

Reply:
Took a look at this and it seems like there's not a clean way around this. We can revisit in future when looking at the `PathResolver trait more generally.